File: my-solutions/0015-3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        
        for i, a in enumerate(nums):
            # skip through all elements that are same from the left
            if i > 0 and a == nums[i-1]:
                continue
                
            l, r = i + 1, len(nums) - 1
            while(l < r):
                sum = nums[l] + nums[r] + a
                # print([nums[l], nums[r], nums[i]])
                if sum == 0:
                    res.append([nums[l], nums[r], nums[i]])
                    
                    l += 1
                    while nums[l] == nums[l-1] and l < r:
                        l += 1
                elif sum > 0:
                    r -= 1
                else:
                    l += 1
        
        return res

File: my-solutions/0015-3sum/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for(int i = 0; i < nums.size() - 2; ++i) {
            if(i > 0 && nums[i] == nums[i-1]) continue;
            int l = i + 1, r = nums.size() - 1;
            while(l < r) {
                int sum = nums[i] + nums[l] + nums[r]; // Store sum for clarity and efficiency
                if(sum == 0) {
                    res.push_back({nums[i], nums[l], nums[r]});
                    ++l;
                    while(l < r && nums[l] == nums[l - 1]) ++l;
                    --r;
                    while(l < r && nums[r] == nums[r + 1]) --r;
                }
                else if(sum > 0) --r;
                else ++l;
            }
        }
        return res;
    }
};
